core/src/ledger/storage/mod.rs

```diff
@@ -1182,6 +1193,52 @@ where
             .unwrap_or_default()
     }
 
+    /// Get the oldest epoch where we can read a value
+    pub fn get_oldest_epoch(&self) -> Epoch {
+        let oldest_height = match self.storage_read_past_height_limit {
+            Some(limit) if limit < self.get_last_block_height().0 => {
+                (self.get_last_block_height().0 - limit).into()
+            }
+            _ => BlockHeight(1),
+        };
+        self.block
+            .pred_epochs
+            .get_epoch(oldest_height)
+            .unwrap_or_default()
+    }
+
+    /// Get oldest epoch which has the valid signed nonce of the bridge pool
+    pub fn get_oldest_epoch_with_valid_nonce(&self) -> Result<Epoch> {
+        let nonce_key = bridge_pool::get_nonce_key();
+        let (bytes, _) = self.read(&nonce_key)?;
+        let bytes = bytes.expect("Bridge pool nonce should exits");
+        let current_nonce =
+            Uint::try_from_slice(&bytes).map_err(Error::BorshCodingError)?;
+        let (mut epoch, _) = self.get_last_epoch();
+        // We don't need to check the older epochs because their Merkle tree
+        // snapshots have been already removed
+        let oldest_epoch = self.get_oldest_epoch();
+        // Look up the last valid epoch which has the previous nonce of the
+        // current one. It has the previous nonce, but it was
+        // incremented during the epoch.
+        while 0 < epoch.0 && oldest_epoch <= epoch {
+            epoch = epoch.prev();
+            let height =
+                match self.block.pred_epochs.get_start_height_of_epoch(epoch) {
+                    Some(h) => h,
+                    None => continue,
+                };
+            let (bytes, _) = self.read_with_height(&nonce_key, height)?;
+            let bytes = bytes.expect("Bridge pool nonce should exits");
+            let nonce = Uint::try_from_slice(&bytes)
+                .map_err(Error::BorshCodingError)?;
+            if nonce < current_nonce {
+                break;
```

Comment:
Isn't this returning at the first epoch with an invalid nonce instead of the last one matching the current nonce?

Reply:
The nonce was incremented to the current one during the epoch though the epoch has the invalid nonce at the starting height. The Merkle tree at the epoch starting would be required to rebuild one at a height (after the nonce was incremented) during the epoch. So, the Merkle tree in the epoch should have a valid nonce.
There is a corner case; the nonce was incremented at the same height when the epoch was updated. However, the Merkle tree snapshot would be pruned eventually.